cupy/random/cupy_distributions.cu

```diff
@@ -164,6 +164,61 @@ __device__ double rk_beta(rk_state* state, double a, double b) {
     }
 }
 
+__device__ long rk_poisson_mult(rk_state *state, double lam) {
+    long X;
+    double prod, U, enlam;
+    enlam = exp(-lam);
+    X = 0;
+    prod = 1.0;
+    while (1) {
+        U = state->rk_double();
+        prod *= U;
+        if (prod > enlam) {
+            X += 1;
+        } else {
+            return X;
+        }
+    }
+}
+
+__device__ long rk_poisson_ptrs(rk_state *state, double lam) {
+    long k;
+    double U, V, slam, loglam, a, b, invalpha, vr, us;
+    slam = sqrt(lam);
+    loglam = log(lam);
+    b = 0.931 + 2.53*slam;
+    a = -0.059 + 0.02483*b;
+    invalpha = 1.1239 + 1.1328/(b-3.4);
+    vr = 0.9277 - 3.6224/(b-2);
+    while (1) {
+        U = state->rk_double(); - 0.5;
+        V = state->rk_double();
+        us = 0.5 - fabs(U);
+        k = (long)floor((2*a/us + b)*U + lam + 0.43);
+        if ((us >= 0.07) && (V <= vr)) {
+            return k;
+        }
+        if ((k < 0) ||
+            ((us < 0.013) && (V > us))) {
+            continue;
+        }
+        if ((log(V) + log(invalpha) - log(a/(us*us)+b)) <=
+            (-lam + k*loglam - loggam(k+1))) {
+            return k;
+        }
+    }
+}
+
+__device__ long rk_poisson(rk_state *state, double lam) {
```

Comment:
is the return type long correct here? I think it should be double

Reply:
I use long because according to the NumPy docs 'output is limited to the range of the C int64 type' and also the current implementation of poisson distribution in RandomState module uses long as a return type.

Reply:
If NumPy is right on the `int64` return type, then this must be `long long`, not `long`, because on Windows a `long` is 32 bits, not 64. See LP64 vs LLP64: https://docs.nvidia.com/cuda/nvrtc/index.html#integer-size.

Reply:
(I didn't read the code here, just wanna point out `long` is not portable. So don't mind me if this is irrelevant of the implementation here. It's just that I spent quite some time fixing Windows errors, so wanna make sure we get it right from now on 😁)

Reply:
let's use `int64_t`

Reply:
I just confirmed that the datatype in windows is `int64`